File: Flatshare Code/FlatShare 2.0/FlatshareCentral/src/main/java/com/joinflatshare/ui/profile/edit/ProfileEditImageAdapter.java
package com.joinflatshare.ui.profile.edit;

import static com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade;

import android.Manifest;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.Target;
import com.google.android.material.card.MaterialCardView;
import com.joinflatshare.FlatshareCentral.R;
import com.joinflatshare.customviews.bottomsheet.BottomSheetView;
import com.joinflatshare.customviews.bottomsheet.ModelBottomSheet;
import com.joinflatshare.interfaces.OnitemClick;
import com.joinflatshare.ui.base.BaseActivity;
import com.joinflatshare.utils.helper.CommonMethod;
import com.joinflatshare.utils.helper.CommonMethods;
import com.joinflatshare.utils.helper.CommonMethod;
import com.joinflatshare.utils.helper.ImageHelper;
import com.joinflatshare.utils.permission.PermissionUtil;

import java.util.ArrayList;

public class ProfileEditImageAdapter extends RecyclerView.Adapter<ProfileEditImageAdapter.ViewHolder> {
    private ArrayList<String> items = new ArrayList<>();
    private final BaseActivity activity;
    private int width, height;
    private BottomSheetView bottomSheetView;

    private OnitemClick onitemClick;

    public ProfileEditImageAdapter(BaseActivity activity) {
        this.activity = activity;
        init();
    }

    public ProfileEditImageAdapter(BaseActivity activity, OnitemClick onitemClick) {
        this.onitemClick = onitemClick;
        this.activity = activity;
        init();
    }

    private void init() {
        if (activity instanceof ProfileEditActivity)
            prepareBottomMenu();
        width = (CommonMethods.getScreenWidth() / 2) - 30;
        height = (width / 3) * 4;
    }

    private void prepareBottomMenu() {
        ArrayList<ModelBottomSheet> menu = new ArrayList<>();
        menu.add(new ModelBottomSheet(R.drawable.ic_cross_red, "Remove Photo"));
        bottomSheetView = new BottomSheetView(activity, menu);
    }

    public void setItems(ArrayList<String> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    public ArrayList<String> getItems() {
        return items;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_profile_image, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(width, height);
        holder.img_profile_banner1.setLayoutParams(params);

        holder.img_profile_banner1.setImageResource(0);
        holder.img_profile_add1.setVisibility(View.VISIBLE);

        String item = items.get(holder.getBindingAdapterPosition());
        holder.img_profile_banner1.setImageResource(0);
        holder.img_profile_add1.setVisibility(View.VISIBLE);
        if (!item.isEmpty()) {
            holder.img_profile_add1.setVisibility(View.GONE);
            if (item.startsWith("Images/") && items.size() > holder.getBindingAdapterPosition())
                loadImage(holder.getBindingAdapterPosition(), ImageHelper.getUserImagesWithAws(item), holder);
            else holder.img_profile_banner1.setImageURI(Uri.parse(item));
        }

        holder.frame_image.setOnClickListener(v -> {
            if (activity instanceof ProfileEditActivity) {
                final ProfileEditActivity act = (ProfileEditActivity) activity;
                CommonMethod.INSTANCE.makeLog("Position", "" + holder.getBindingAdapterPosition());
                if (!item.isEmpty()) {
                    bottomSheetView.show((view, position1) -> {
                        if (item.startsWith("Images/")) {
                            act.dataBind.deletedUserImages.add(item);
                        }
                        act.dataBind.addedUserImages.remove(act.dataBind.adapterUserImages.get(holder.getBindingAdapterPosition()));
                        act.dataBind.adapterUserImages.remove(holder.getBindingAdapterPosition());
                        if (act.dataBind.adapterUserImages.size() < 10 && !act.dataBind.adapterUserImages.get(0).isEmpty())
                            act.dataBind.adapterUserImages.add(0, "");
                        notifyDataSetChanged();
                        act.dataBind.setCompleteCount();
                    });
                } else {
                    act.imageClickPosition = holder.getBindingAdapterPosition();
                    PermissionUtil.INSTANCE.validatePermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE, granted -> {
                        if (granted)
                            act.pickImage();
                        else CommonMethod.INSTANCE.makeToast("Permission not provided");
                    });
                }
            } else {
                if (onitemClick != null) {
                    onitemClick.onitemclick(holder.itemView, holder.getBindingAdapterPosition());
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        ImageView img_profile_banner1, img_profile_add1;
        MaterialCardView frame_image;

        ViewHolder(View itemView) {
            super(itemView);
            img_profile_banner1 = itemView.findViewById(R.id.img_profile_banner1);
            img_profile_add1 = itemView.findViewById(R.id.img_profile_add1);
            frame_image = itemView.findViewById(R.id.frame_image);
        }
    }

    private void loadImage(int position, String url, ViewHolder holder) {
        Glide.with(activity)
                .load(url)
                .apply(
                        new RequestOptions()
                                .error(0)
                                .centerCrop()
                )
                .listener(new RequestListener<Drawable>() {
                    @Override
                    public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                        holder.img_profile_banner1.setImageResource(0);
                        holder.img_profile_add1.setVisibility(View.VISIBLE);
                        items.set(position, "");
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                        CommonMethod.INSTANCE.makeLog("Image URL", url);
                        holder.img_profile_banner1.setImageDrawable(resource);
                        holder.img_profile_add1.setVisibility(View.GONE);
                        return false;
                    }
                })
                .transition(withCrossFade())
                .into(holder.img_profile_banner1);
    }
}

File: Flatshare Code/FlatShare 2.0/FlatshareCentral/src/main/java/com/joinflatshare/ui/profile/edit/ProfileEditImageAdapter.kt
package com.joinflatshare.ui.profile.edit

import android.Manifest
import android.graphics.drawable.Drawable
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.Target
import com.joinflatshare.FlatshareCentral.R
import com.joinflatshare.FlatshareCentral.databinding.ItemProfileImageBinding
import com.joinflatshare.customviews.bottomsheet.BottomSheetView
import com.joinflatshare.customviews.bottomsheet.ModelBottomSheet
import com.joinflatshare.utils.helper.CommonMethod.makeLog
import com.joinflatshare.utils.helper.CommonMethod.makeToast
import com.joinflatshare.utils.helper.ImageHelper
import com.joinflatshare.utils.permission.PermissionUtil

class ProfileEditImageAdapter(
    private val activity: ProfileEditActivity
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var height = 0
    private val items = ArrayList<String>()

    fun setItems(items: ArrayList<String>) {
        this.items.clear()
        this.items.addAll(items)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val viewBind =
            ItemProfileImageBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(viewBind)
    }

    override fun onBindViewHolder(mainHolder: RecyclerView.ViewHolder, position: Int) {
        val holder = mainHolder as ViewHolder
        holder.bind(mainHolder.adapterPosition, this)
    }

    override fun getItemCount(): Int {
        return items.size
    }

    class ViewHolder(
        private val view: ItemProfileImageBinding
    ) : RecyclerView.ViewHolder(view.root) {
        fun bind(
            position: Int, adapter: ProfileEditImageAdapter
        ) {
            if (position == 0 && adapter.height == 0) {
                view.frameImage.viewTreeObserver.addOnGlobalLayoutListener(object :
                    OnGlobalLayoutListener {
                    override fun onGlobalLayout() {
                        view.frameImage.viewTreeObserver.removeOnGlobalLayoutListener(this)
                        val width = view.frameImage.width
                        adapter.height = (width / 3) * 4
                        adapter.notifyDataSetChanged()
                    }
                })
            } else {
                view.frameImage.layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    adapter.height
                )
                view.imgProfileAdd1.visibility = View.VISIBLE
                view.imgProfileBanner1.setImageResource(0)

                val item = adapter.items[position]
                if (item.isNotEmpty()) {
                    view.imgProfileAdd1.setVisibility(View.GONE)
                    if (item.startsWith("Images/"))
                        loadImage(
                            adapter,
                            position, ImageHelper.getUserImagesWithAws(item), view
                        )
                    else view.imgProfileBanner1.setImageURI(Uri.parse(item))
                }
                view.frameImage.setOnClickListener { _: View? ->
                    makeLog("Position", "" + position)
                    if (item.isNotEmpty()) {
                        val menu = ArrayList<ModelBottomSheet>()
                        menu.add(ModelBottomSheet(R.drawable.ic_cross_red, "Remove Photo"))
                        BottomSheetView(
                            adapter.activity, menu
                        ) { _, position ->
                            if (item.startsWith("Images/")) {
                                adapter.activity.dataBind.deletedUserImages.add(item)
                            }
                            adapter.activity.dataBind.addedUserImages.remove(
                                adapter.activity.dataBind.adapterUserImages[position]
                            )
                            adapter.activity.dataBind.adapterUserImages.removeAt(position)
                            adapter.notifyDataSetChanged()
                        }
                    } else {
                        adapter.activity.imageClickPosition = position
                        PermissionUtil.validatePermission(
                            adapter.activity,
                            Manifest.permission.READ_EXTERNAL_STORAGE
                        ) { granted: Boolean ->
                            if (granted) adapter.activity.pickImage()
                            else makeToast("Permission not provided")
                        }
                    }
                    /*else {
                       val stfalconViewer: StfalconImageViewer<String> =
                           StfalconImageViewer.Builder(
                               adapter.activity, adapter.items
                           ) { imageView: ImageView?, imageUrl: String? ->
                               ImageHelper.loadImage(
                                   adapter.activity,
                                   R.mipmap.ic_launcher,
                                   imageView,
                                   imageUrl
                               )
                           }.allowZooming(true).withBackgroundColor(
                               ContextCompat.getColor(
                                   adapter.activity,
                                   R.color.color_bg
                               )
                           ).withHiddenStatusBar(true).withStartPosition(position).build()
                       stfalconViewer.show()
                   }*/
                }
            }
        }


        private fun loadImage(
            adapter: ProfileEditImageAdapter,
            position: Int,
            url: String,
            holder: ItemProfileImageBinding
        ) {
            Glide.with(adapter.activity)
                .load(url)
                .apply(
                    RequestOptions()
                        .error(0)
                        .centerCrop()
                )
                .listener(object : RequestListener<Drawable?> {
                    override fun onLoadFailed(
                        e: GlideException?,
                        model: Any?,
                        target: Target<Drawable?>,
                        isFirstResource: Boolean
                    ): Boolean {
                        holder.imgProfileBanner1.setImageResource(0)
                        holder.imgProfileAdd1.setVisibility(View.VISIBLE)
                        adapter.items[position] = ""
                        return false
                    }

                    override fun onResourceReady(
                        resource: Drawable,
                        model: Any,
                        target: Target<Drawable?>?,
                        dataSource: DataSource,
                        isFirstResource: Boolean
                    ): Boolean {
                        makeLog("Image URL", url)
                        holder.imgProfileBanner1.setImageDrawable(resource)
                        holder.imgProfileAdd1.setVisibility(View.GONE)
                        return false
                    }
                })
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(holder.imgProfileBanner1)
        }
    }


}